### PR TITLE
Settings export

### PR DESCRIPTION
**Pull request by Somethingweirdhere**
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

Alright, this actually seems to work now! I haven't encountered any issues with this yet, but I will continue testing it because exports feel quite important. I hope someone else can also confirm there are no issues.

Anyways, I had to change how ZipHelper.extractFileFromZip works because it was broken and always returned true and also didn't work for the same ZipInputStream more than once. It now uses the filepath, but as the function isn't used outside of ContentSettingsFragment.java this won't be an issue. 

The exports now contain a new file called newpipe.settings. If while importing it is detected, it will trigger a second dialog asking if to import settings too. If you press Ok it will import the settings from there.

#849

### CODE REVIEW

**Comment by Somethingweirdhere** [comment]
@theScrabi Is everything alright with this or should I change something with the way it works?

**Comment by theScrabi** [comment]
@Somethingweird I am sorry I didn't find time to check everything yet.
I hope I can do a review at the weekend.

File: app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java [hunk 1]
```diff
@@ -207,6 +220,29 @@ private void exportDatabase(String path) {
         }
     }
 
+    private void saveSharedPreferencesToFile(File dst) {
```
Comment (theScrabi) [inline app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java]: Why can't we save `/shared_prefs/org.schabi.newpipe.debug_preferences.xml` directly? Is there any advantage generating by generically?

File: app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java [hunk 2]
```diff
@@ -223,30 +259,89 @@ private void importDatabase(String filePath) {
...
 
-            // restart app to properly load db
-            //App.restart(getContext());
-            System.exit(0);
         } catch (Exception e) {
             onError(e);
         }
     }
 
+    private void loadSharedPreferences(File src) {
```
Comment (theScrabi) [inline app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java]: Same here.

File: app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java [hunk 3]
```diff
@@ -223,30 +259,89 @@ private void importDatabase(String filePath) {
...
 
-            zipIn.close();
+            //If settings file exist, ask if it should be imported.
+            if(ZipHelper.extractFileFromZip(filePath, newpipe_settings.getPath(), "newpipe.settings")) {
+                AlertDialog.Builder alert = new AlertDialog.Builder(getContext());
+                alert.setTitle(R.string.import_settings);
+
+                alert.setNegativeButton(android.R.string.no, new DialogInterface.OnClickListener() {
+                    @Override
+                    public void onClick(DialogInterface dialog, int which) {
```
Comment (theScrabi) [inline app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java]: Please use lambda convention here.

File: app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java [hunk 4]
```diff
@@ -223,30 +259,89 @@ private void importDatabase(String filePath) {
...
+                        dialog.dismiss();
+                        // restart app to properly load db
+                        System.exit(0);
+                    }
+                });
+                alert.setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
+                    @Override
+                    public void onClick(DialogInterface dialog, int which) {
```
Comment (theScrabi) [inline app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java]: Please use lambda convention here.

**Comment by Somethingweirdhere** [comment]
@theScrabi 

I changed to lambda convention. I am not sure why travis failed, I hope it's nothing serious.

Now about the saving:

I don't see any disadvantages to the way I did it, are there any except that there is more code? I did it this way because of some very minor advantages, but I tested it and it also works like you suggested.

The advantages are:

1. Exporting the preferences one by one and reimporting means that even if the way how preferences work changes (like the xml header, the file where it is stored, or if there is more information in the file than now exists), the exporting/importing will continue working. I don't see why this would change, but it's a possibility I guess.

2. [https://stackoverflow.com/questions/6146106/where-are-shared-preferences-stored/](Here) it is mentioned in the comments of the answers that some devices store their DefaultPreferences in differently names files. I haven't been able to find a device that does this, but I only have a S8+ and the provided emulators, but those only work up from Lollipop for me. I also haven't been able to find information about this topic anywhere else. It seems unlikely to me that there is a device that does not store the preferences in org.schabi.newpipe.debug_preferences.xml, but I didn't want to risk it.

It is up to you which way you prefer, changing the code won't take me long.

**Comment by theScrabi** [comment]
muy bien.

**Comment by theScrabi** [comment]
Don't forget to claim your reward at bountysource:
https://www.bountysource.com/issues/51556592-feature-request-ability-to-import-export-settings

**Pull request merged**